### ISSUE TITLE
`findnext(A, start)` and `findnext(testf::Function, A, start)` behave differently

### ISSUE THREAD

**Issue by goretkin** [JuliaLang/julia#36963]
e.g.

```julia
using InfiniteArrays

# works
findfirst(>(5), 1:InfiniteArrays.Infinity())

# does not
findfirst(1:InfiniteArrays.Infinity() .> 5)
```

Divergence at https://github.com/JuliaLang/julia/pull/35965

**Comment by dlfivefifty** [comment]
Just needs

https://github.com/JuliaLang/julia/blob/dffc8895b897a78a8dca64c784962675df49c344/base/array.jl#L1725

To become
```julia
i = oftype(first(keys(A)), start)
```
Shall I make a PR?

**Comment by nalimilan** [comment]
Probably better convert to `eltype(keys(A))` instead, to avoid special-casing the last index.

**Comment by goretkin** [comment]
@dlfivefifty I would prefer to get #36965 in as a solution. And then we could follow @nalimilan's suggestion, and we'll only have to fix it in half as many places.

**Comment by vtjnash** [comment]
Addressed by #36965, IIUC
